### PR TITLE
+str #19593 Fix a potential race condition in FileSubscriber.

### LINKED ISSUE #19593: Potential race condition in FileSubscriber
When Akka Stream `FileSubscriber` completes with an error, the `Promise` is completed before closing the channel (and file). This will generate a race condition where a client wanting to delete the file upon the promise completion will fail (on some systems) because the file is still open.

Solution is to _always_ close the channel before completing the promise.

### PR DESCRIPTION
**Pull request by ochrons**
If an error occurs in the stream, the completionPromise is completed
before the file is closed, leading to potential race condition if the
file is then tried to be deleted upon the promise completion, while it's
still open.

Fixes #19593

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala [hunk 1]
```diff
@@ -46,7 +46,7 @@ private[akka] class FileSubscriber(f: File, completionPromise: Promise[IOResult]
     super.preStart()
   } catch {
     case ex: Exception ⇒
-      completionPromise.success(IOResult(bytesWritten, Failure(ex)))
+      closeAndComplete(IOResult(bytesWritten, Failure(ex)))
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala]: how about a finally block?

Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala]: `Nonfatal`?

**Comment by drewhk** [comment]
OK TO TEST

**Comment by drewhk** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala [hunk 2]
```diff
@@ -70,20 +70,26 @@ private[akka] class FileSubscriber(f: File, completionPromise: Promise[IOResult]
...
+    closeAndComplete(IOResult(bytesWritten, Success(Done)))
+    super.postStop()
+  }
+
+  private def closeAndComplete(result: IOResult): Unit = {
     try {
+      // close the channel/file before completing the promise, allowing the
+      // file to be deleted, which would not work (on some systems) if the
+      // file is still open for writing
       if (chan ne null) chan.close()
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala]: perhaps check  `chan.isOpen`?

Comment (ochrons) [inline akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala]: "If the channel has already been closed then this method returns immediately.", so no need to do that check ourselves.

Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala]: Agreed :)

**Comment by johanandren** [comment]
LGTM

**Pull request merged**